qcodes/dataset/data_set_protocol.py

```diff
@@ -0,0 +1,155 @@
+from __future__ import annotations
+
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Sized,
+    Tuple,
+    Union,
+)
+
+import numpy as np
+from typing_extensions import Protocol, runtime_checkable
+
+from qcodes.dataset.descriptions.dependencies import InterDependencies_
+from qcodes.dataset.descriptions.param_spec import ParamSpec, ParamSpecBase
+from qcodes.dataset.descriptions.rundescriber import RunDescriber
+from qcodes.dataset.descriptions.versioning.rundescribertypes import Shapes
+from qcodes.dataset.export_config import DataExportType
+from qcodes.dataset.linked_datasets.links import Link
+
+SPECS = List[ParamSpec]
+# Transition period type: SpecsOrInterDeps. We will allow both as input to
+# the DataSet constructor for a while, then deprecate SPECS and finally remove
+# the ParamSpec class
+SpecsOrInterDeps = Union[SPECS, InterDependencies_]
+
+
+if TYPE_CHECKING:
+
+    from .data_set_cache import DataSetCache
+
+
+@runtime_checkable
+class DataSetProtocol(Protocol, Sized):
+    def prepare(
+        self,
+        *,
+        snapshot: Mapping[Any, Any],
```

Comment:
what i'm not sure is the snapshot. so `prepare` takes a dict, while `add_snapshot` takes a string, and the only declared way to get the snapshot is `snapshot_raw` which returns a string, not a dict. I don't feel good about these inconsistencies, and I am not sure what's the best to do. idea 1 is to let datasetprotocol in the first instance only take snapshot as a string (so change prepare to take a string) thus making dataset competely agnostic to the content/structure/serialization of the snapshot. idea 2 is to keep things as they are but also add the `snapshot` property from current DataSet that returns snapshot that's deserialized from string, this forces all datasetprotocol implementers to provide that interface that users actually use (noone cares about `snapshot_raw`). perhaps idea 1 is the simplest for this PR, and idea 2 can be implemented later if needed. what do you think @jenshnielsen ?

Reply:
I think it would be fine to add the snapshot property back to the protocol. I have initially aimed at removing as much as possible. The other question is if we should make some of these methods and properties private. Just because `snapshot_raw` is public on the DataSet class it needs not be public on the interface. The public method would then simply be implemented as a wrapper around the private one

Reply:
OK, I have added snapshot as a property and made snapshot_raw private in the protocol (retaining it in the dataset class for backwards compatibility)